File: cache/cache-annotation-processor/src/main/java/ru/tinkoff/kora/cache/annotation/processor/aop/CacheableAopKoraAspect.java
package ru.tinkoff.kora.cache.annotation.processor.aop;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import reactor.core.publisher.Flux;
import reactor.core.scheduler.Schedulers;
import ru.tinkoff.kora.annotation.processor.common.MethodUtils;
import ru.tinkoff.kora.annotation.processor.common.ProcessingErrorException;
import ru.tinkoff.kora.cache.annotation.processor.CacheOperation;
import ru.tinkoff.kora.cache.annotation.processor.CacheOperationUtils;

import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ExecutableElement;
import java.util.List;
import java.util.Set;
import java.util.concurrent.Future;

public class CacheableAopKoraAspect extends AbstractAopCacheAspect {

    private static final ClassName ANNOTATION_CACHEABLE = ClassName.get("ru.tinkoff.kora.cache.annotation", "Cacheable");
    private static final ClassName ANNOTATION_CACHEABLES = ClassName.get("ru.tinkoff.kora.cache.annotation", "Cacheables");

    private final ProcessingEnvironment env;

    public CacheableAopKoraAspect(ProcessingEnvironment env) {
        this.env = env;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Set.of(ANNOTATION_CACHEABLE.canonicalName(), ANNOTATION_CACHEABLES.canonicalName());
    }

    @Override
    public ApplyResult apply(ExecutableElement method, String superCall, AspectContext aspectContext) {
        if (MethodUtils.isFuture(method)) {
            throw new ProcessingErrorException("@Cacheable can't be applied for types assignable from " + Future.class, method);
        } else if (MethodUtils.isFlux(method)) {
            throw new ProcessingErrorException("@Cacheable can't be applied for types assignable from " + Flux.class, method);
        }

        final CacheOperation operation = CacheOperationUtils.getCacheMeta(method);
        final List<String> cacheFields = getCacheFields(operation, env, aspectContext);

        final CodeBlock body = MethodUtils.isMono(method)
            ? buildBodyMono(method, operation, cacheFields, superCall)
            : buildBodySync(method, operation, cacheFields, superCall);

        return new ApplyResult.MethodBody(body);
    }

    private CodeBlock buildBodySync(ExecutableElement method,
                                    CacheOperation operation,
                                    List<String> cacheFields,
                                    String superCall) {
        final String superMethod = getSuperMethod(method, superCall);

        final CodeBlock keyBlock;
        if (operation.parameters().size() == 1) {
            keyBlock = CodeBlock.builder()
                .add("""
                        var _key = $L;
                        """,
                    operation.parameters().get(0))
                .build();
        } else {
            final String recordParameters = getKeyRecordParameters(operation, method);
            keyBlock = CodeBlock.builder()
                .add("""
                        var _key = $T.of($L);
                        """,
                    getCacheKey(operation), recordParameters)
                .build();

        }

        if (operation.cacheImplementations().size() == 1) {
            return CodeBlock.builder()
                .add(keyBlock)
                .add(CodeBlock.of("""
                    return $L.computeIfAbsent(_key, _k -> $L);
                    """, cacheFields.get(0), superMethod))
                .build();
        }

        final StringBuilder builder = new StringBuilder();

        // cache get
        for (int i = 0; i < cacheFields.size(); i++) {
            final String cache = cacheFields.get(i);
            final String prefix = (i == 0)
                ? "var _value = "
                : "_value = ";

            builder.append(prefix)
                .append(cache).append(".get(_key);\n")
                .append("if(_value != null) {\n");

            // put value from cache into prev level caches
            for (int j = 0; j < i; j++) {
                final String cachePrevPut = cacheFields.get(j);
                builder.append("\t").append(cachePrevPut).append(".put(_key, _value);\n");
            }

            builder.append("""
                        return _value;
                     }
                    """)
                .append("\n");
        }

        // cache super method
        builder.append("_value = ").append(superMethod).append(";\n");

        // cache put
        for (final String cache : cacheFields) {
            builder.append(cache).append(".put(_key, _value);\n");
        }
        builder.append("return _value;");

        return CodeBlock.builder()
            .add(keyBlock)
            .add(builder.toString())
            .build();
    }

    private CodeBlock buildBodyMono(ExecutableElement method,
                                    CacheOperation operation,
                                    List<String> cacheFields,
                                    String superCall) {
        final String superMethod = getSuperMethod(method, superCall);

        // cache variables
        final CodeBlock.Builder builder = CodeBlock.builder();

        // cache get
        for (int i = 0; i < cacheFields.size(); i++) {
            final String cache = cacheFields.get(i);
            final String prefix = (i == 0)
                ? "var _value = "
                : "_value = _value.switchIfEmpty(";

            final String getPart = ".getAsync(_key)";
            builder.add(prefix)
                .add(cache)
                .add(getPart);

            // put value from cache into prev level caches
            if (i > 1) {
                builder.add("\n").add("""
                        .publishOn($T.boundedElastic())
                        .doOnSuccess(_fromCache -> {
                            if(_fromCache != null) {
                                $T.merge($T.of(
                    """, Schedulers.class, Flux.class, List.class);

                for (int j = 0; j < i; j++) {
                    final String prevCache = cacheFields.get(j);
                    final String suffix = (j == i - 1)
                        ? ".putAsync(_key, _fromCache)\n"
                        : ".putAsync(_key, _fromCache),\n";
                    builder.add("\t\t\t\t").add(prevCache).add(suffix);
                }

                builder.add("\t\t)).then().block();\n}}));\n\n");
            } else if (i == 1) {
                builder.add("\n\t")
                    .add(String.format("""
                        .doOnSuccess(_fromCache -> {
                                if(_fromCache != null) {
                                    %s.put(_key, _fromCache);
                                }
                        }));
                        """, cacheFields.get(0)))
                    .add("\n");
            } else {
                builder.add(";\n");
            }
        }

        // cache super method
        builder.add("return _value.switchIfEmpty(").add(superMethod);

        // cache put
        if (cacheFields.size() > 1) {
            builder.add(".flatMap(_result -> $T.merge($T.of(\n", Flux.class, List.class);
            for (int i = 0; i < cacheFields.size(); i++) {
                final String cache = cacheFields.get(i);
                final String suffix = (i == cacheFields.size() - 1)
                    ? ".putAsync(_key, _result)\n"
                    : ".putAsync(_key, _result),\n";
                builder.add("\t").add(cache).add(suffix);
            }
            builder.add(")).then(Mono.just(_result))));");
        } else {
            if (operation.parameters().size() == 1) {
                builder.add("""
                    .doOnSuccess(_result -> {
                        if(_result != null) {
                            $L.put($L, _result);
                        }
                    }));
                    """, cacheFields.get(0), operation.parameters().get(0));
            } else {
                final String recordParameters = getKeyRecordParameters(operation, method);
                builder.add("""
                    .doOnSuccess(_result -> {
                        if(_result != null) {
                            $L.put($T.of($L), _result);
                        }
                    }));
                    """, cacheFields.get(0), getCacheKey(operation), recordParameters);
            }
        }

        if (operation.parameters().size() == 1) {
            return CodeBlock.builder()
                .add("""
                        var _key = $L;
                        """,
                    operation.parameters().get(0))
                .add(builder.build())
                .build();
        } else {
            final String recordParameters = getKeyRecordParameters(operation, method);
            return CodeBlock.builder()
                .add("""
                        var _key = $T.of($L);
                        """,
                    getCacheKey(operation), recordParameters)
                .add(builder.build())
                .build();

        }
    }
}

File: cache/cache-symbol-processor/src/main/kotlin/ru/tinkoff/kora/cache/symbol/processor/aop/CacheableAopKoraAspect.kt
package ru.tinkoff.kora.cache.symbol.processor.aop

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ksp.toClassName
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ru.tinkoff.kora.aop.symbol.processor.KoraAspect
import ru.tinkoff.kora.cache.symbol.processor.CacheOperation
import ru.tinkoff.kora.cache.symbol.processor.CacheOperationUtils.Companion.getCacheOperation
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFlux
import ru.tinkoff.kora.ksp.common.FunctionUtils.isFuture
import ru.tinkoff.kora.ksp.common.FunctionUtils.isMono
import ru.tinkoff.kora.ksp.common.FunctionUtils.isSuspend
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.concurrent.Future

@KspExperimental
class CacheableAopKoraAspect(private val resolver: Resolver) : AbstractAopCacheAspect() {

    private val CAFFEINE_CACHE = ClassName("ru.tinkoff.kora.cache.caffeine", "CaffeineCache")
    private val ANNOTATION_CACHEABLE = ClassName("ru.tinkoff.kora.cache.annotation", "Cacheable")
    private val ANNOTATION_CACHEABLES = ClassName("ru.tinkoff.kora.cache.annotation", "Cacheables")

    override fun getSupportedAnnotationTypes(): Set<String> {
        return setOf(ANNOTATION_CACHEABLE.canonicalName, ANNOTATION_CACHEABLES.canonicalName)
    }

    override fun apply(method: KSFunctionDeclaration, superCall: String, aspectContext: KoraAspect.AspectContext): KoraAspect.ApplyResult {
        if (method.isFuture()) {
            throw ProcessingErrorException("@Cacheable can't be applied for types assignable from ${Future::class.java}", method)
        } else if (method.isMono()) {
            throw ProcessingErrorException("@Cacheable can't be applied for types assignable from ${Mono::class.java}", method)
        } else if (method.isFlux()) {
            throw ProcessingErrorException("@Cacheable can't be applied for types assignable from ${Flux::class.java}", method)
        }

        val operation = getCacheOperation(method)
        val cacheFields = getCacheFields(operation, resolver, aspectContext)

        val body = if (method.isSuspend()) {
            buildBodySync(method, operation, superCall, cacheFields, resolver)
        } else {
            buildBodySync(method, operation, superCall, cacheFields, resolver)
        }

        return KoraAspect.ApplyResult.MethodBody(body)
    }

    private fun buildBodySync(
        method: KSFunctionDeclaration,
        operation: CacheOperation,
        superCall: String,
        cacheFields: List<String>,
        resolver: Resolver
    ): CodeBlock {
        val recordParameters = getKeyRecordParameters(operation, method)
        val superMethod = getSuperMethod(method, superCall)
        val builder = CodeBlock.builder()
        val isSingleNullableParam = operation.parameters.size == 1 && operation.parameters[0].type.resolve().isMarkedNullable

        val keyBlock = if (operation.parameters.size == 1) {
            CodeBlock.of("val _key = %L\n", operation.parameters[0])
        } else {
            CodeBlock.of("val _key = %T.of(%L)\n", getCacheKey(operation), recordParameters)
        }

        if (!method.isSuspend() && operation.cacheImplementations.size == 1) {
            val impl = resolver.getClassDeclarationByName(operation.cacheImplementations[0])
            if (impl != null) {
                val codeBlock = if (isSingleNullableParam) {
                    CodeBlock.of(
                        """
                                return if (_key != null) {
                                    %L.computeIfAbsent(_key) { %L }
                                } else {
                                    %L
                                }
                            """.trimIndent(), cacheFields[0], superMethod, superMethod
                    )
                } else {
                    CodeBlock.of("return %L.computeIfAbsent(_key) { %L }", cacheFields[0], superMethod)
                }

                return CodeBlock.builder()
                    .add(keyBlock)
                    .add(codeBlock)
                    .build()
            }
        }

        // cache get
        for (i in cacheFields.indices) {
            val cache = cacheFields[i]
            val prefix = if (i == 0) "var _value = " else "_value = "

            if (isSingleNullableParam) {
                builder.add(prefix)
                    .add("_key?.let { %L.get(it) }\n", cache)
                    .add("if(_value != null) {\n");
            } else {
                builder.add(prefix)
                    .add("%L.get(_key)\n", cache)
                    .add("if(_value != null) {\n");
            }

            for (j in 0 until i) {
                val prevCache = cacheFields[j]
                builder.add("\t%L.put(_key, _value)\n", prevCache)
            }

            builder
                .add("\treturn _value\n")
                .add("}\n\n")
        }

        // cache super method
        builder.add("_value = %L\n", superMethod)

        // cache put
        for (cache in cacheFields) {
            if (isSingleNullableParam) {
                builder.add("_key?.let { %L.put(it, _value) }\n", cache)
            } else {
                builder.add("%L.put(_key, _value)\n", cache)
            }
        }
        builder.add("return _value")

        return CodeBlock.builder()
            .add(keyBlock)
            .add(builder.build())
            .build()
    }
}
